##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

##  Привет Антон)
#### Мне ведь не кажется, что слишком уж "грязно" написано?!
#### Если не сложно, когда проверишь, напиши пожалуйста, что думаешь, спасибо за внимание )

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('heart_pandas.csv')
data

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0     63    male  non-anginal pain                     145          233   
1     37    male   atypical angina                     130          250   
2     41  female    typical angina                     130          204   
3     56    male    typical angina                     120          236   
4     57  female    typical angina                     120          354   
..   ...     ...               ...                     ...          ...   
298   57  female    typical angina                     140          241   
299   45    male  non-anginal pain                     110          264   
300   68    male    typical angina                     144          193   
301   57    male    typical angina                     130          131   
302   57  female    typical angina                     130          236   

       fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0    greater than 120mg/ml                 normal                      150   
1      lower than 120mg/ml  ST-T wave abnormality                      187   
2      lower than 120mg/ml                 normal                      172   
3      lower than 120mg/ml  ST-T wave abnormality                      178   
4      lower than 120mg/ml  ST-T wave abnormality                      163   
..                     ...                    ...                      ...   
298    lower than 120mg/ml  ST-T wave abnormality                      123   
299    lower than 120mg/ml  ST-T wave abnormality                      132   
300  greater than 120mg/ml  ST-T wave abnormality                      141   
301    lower than 120mg/ml  ST-T wave abnormality                      115   
302    lower than 120mg/ml                 normal                      174   

    exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                        no            2.3  upsloping                  0   
1                        no            3.5  upsloping                  0   
2                        no            1.4       flat                  0   
3                        no            0.8       flat                  0   
4                       yes            0.6       flat                  0   
..                      ...            ...        ...                ...   
298                     yes            0.2  upsloping                  0   
299                      no            1.2  upsloping                  0   
300                      no            3.4  upsloping                  2   
301                     yes            1.2  upsloping                  1   
302                      no            0.0  upsloping                  1   

           thalassemia  target  
0               normal       1  
1         fixed defect       1  
2         fixed defect       1  
3         fixed defect       1  
4         fixed defect       1  
..                 ...     ...  
298  reversable defect       0  
299  reversable defect       0  
300  reversable defect       0  
301  reversable defect       0  
302       fixed defect       0  

[303 rows x 14 columns]

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

#### Сколько мужчин в датасете? Сколько женщен? (sex)

In [16]:
data.sex.value_counts() #data[data.sex=='male'].sex.value_counts()

male      207
female     96
Name: sex, dtype: int64

#### Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) 

In [31]:
data.sex.value_counts(normalize=True).male*100

68.31683168316832

#### Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца?

In [13]:
pd.crosstab(data.target, data.sex)

sex     female  male
target              
0           24   114
1           72    93

**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [7]:
data[data.sex =='male'].target[data.target ==0].count()/data.sex.count()*100

37.62376237623762

**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [53]:
data[data.target == 1].age.min()

29

**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [54]:
data[data.target == 0].age.max()

77

**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [99]:
maf = data.groupby(['sex', 'target'])['age'].min()
maf['female',]

target
0    43
1    34
Name: age, dtype: int64

**Какой средний возраст женщин?**

In [62]:
data[data.sex =='female'].age.mean()

55.677083333333336

**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [33]:
data[data.target == 1].age.mean(), data[data.target == 1].age.std()

(52.4969696969697, 9.550650751946778)

#### Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) 

In [27]:
if data.cholesterol.mean()> data[data.target == 0].cholesterol.mean():
    print('Yes')
else:
    print("No")

No


**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [28]:
df1 = data.groupby('rest_ecg')[['age','resting_blood_pressure',
                          'cholesterol', 'max_heart_rate_achieved', 'num_major_vessels']].mean()
df2 = data.groupby('rest_ecg')[['age','resting_blood_pressure',
                          'cholesterol', 'max_heart_rate_achieved', 'num_major_vessels']].max()

data1 = pd.merge(df1, df2, left_on = 'rest_ecg', right_on='rest_ecg', suffixes=['_mean', '_max'])
data1

age_mean  resting_blood_pressure_mean  \
rest_ecg                                                               
ST-T wave abnormality         52.914474                   129.065789   
left ventricular hypertrophy  61.000000                   140.500000   
normal                        55.687075                   134.027211   

                              cholesterol_mean  max_heart_rate_achieved_mean  \
rest_ecg                                                                       
ST-T wave abnormality               237.269737                    151.960526   
left ventricular hypertrophy        261.750000                    125.750000   
normal                              255.142857                    147.904762   

                              num_major_vessels_mean  age_max  \
rest_ecg                                                        
ST-T wave abnormality                       0.638158       71   
left ventricular hypertrophy                1.000000       76   
normal                                      0.816327       77   

                              resting_blood_pressure_max  cholesterol_max  \
rest_ecg                                                                    
ST-T wave abnormality                                180              354   
left ventricular hypertrophy                         180              327   
normal                                               200              564   

                              max_heart_rate_achieved_max  \
rest_ecg                                                    
ST-T wave abnormality                                 194   
left ventricular hypertrophy                          140   
normal                                                202   

                              num_major_vessels_max  
rest_ecg                                             
ST-T wave abnormality                             4  
left ventricular hypertrophy                      3  
normal                                            4

In [123]:
data.groupby('rest_ecg').describe() ###Я правильно понял?
                                    ### Или имелось ввиду другое решение в одну строку ?

age                                           \
                              count       mean        std   min    25%   50%   
rest_ecg                                                                       
ST-T wave abnormality         152.0  52.914474   9.229196  34.0  44.75  53.5   
left ventricular hypertrophy    4.0  61.000000  10.099505  55.0  55.00  56.5   
normal                        147.0  55.687075   8.675902  29.0  50.50  57.0   

                                         resting_blood_pressure              \
                               75%   max                  count        mean   
rest_ecg                                                                      
ST-T wave abnormality         60.0  71.0                  152.0  129.065789   
left ventricular hypertrophy  62.5  76.0                    4.0  140.500000   
normal                        62.0  77.0                  147.0  134.027211   

                              ... num_major_vessels      target            \
                              ...               75%  max  count      mean   
rest_ecg                      ...                                           
ST-T wave abnormality         ...               1.0  4.0  152.0  0.631579   
left ventricular hypertrophy  ...               1.5  3.0    4.0  0.250000   
normal                        ...               1.0  4.0  147.0  0.462585   

                                                                  
                                   std  min  25%  50%   75%  max  
rest_ecg                                                          
ST-T wave abnormality         0.483971  0.0  0.0  1.0  1.00  1.0  
left ventricular hypertrophy  0.500000  0.0  0.0  0.0  0.25  1.0  
normal                        0.500303  0.0  0.0  0.0  1.00  1.0  

[3 rows x 56 columns]

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [45]:
depression = data.groupby(['sex', 'target'])['st_depression'].mean()
depression
if depression[1]>depression[2]:
    print('male depression is high:', depression[1])
else:
    print('female depression is high:', depression[2])

female depression is high: 1.531578947368421


**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [50]:
gp1 =data.groupby(['chest_pain_type', 'rest_ecg', 'thalassemia'])[['cholesterol']].min()
gp2 = data.groupby(['chest_pain_type', 'rest_ecg', 'thalassemia'])[['cholesterol']].max()
pd.merge(gp1, gp2, left_on=['chest_pain_type', 'rest_ecg', 'thalassemia'],
         right_on=['chest_pain_type', 'rest_ecg', 'thalassemia'], suffixes=['_min', '_max'],  how ='inner')

cholesterol_min  \
chest_pain_type  rest_ecg                     thalassemia                          
atypical angina  ST-T wave abnormality        fixed defect                   141   
                                              normal                         218   
                                              reversable defect              126   
                 left ventricular hypertrophy fixed defect                   197   
                 normal                       fixed defect                   149   
                                              normal                         216   
                                              reversable defect              212   
non-anginal pain ST-T wave abnormality        fixed defect                   204   
                                              reversable defect              199   
                 normal                       fixed defect                   182   
                                              normal                         186   
                                              reversable defect              193   
typical angina   ST-T wave abnormality        fixed defect                   149   
                                              normal                         169   
                                              reversable defect              131   
                 left ventricular hypertrophy fixed defect                   327   
                                              normal                         318   
                                              reversable defect              205   
                 normal                       fixed defect                   166   
                                              normal                         176   
                                              reversable defect              164   

                                                                 cholesterol_max  
chest_pain_type  rest_ecg                     thalassemia                         
atypical angina  ST-T wave abnormality        fixed defect                   340  
                                              normal                         218  
                                              reversable defect              313  
                 left ventricular hypertrophy fixed defect                   197  
                 normal                       fixed defect                   417  
                                              normal                         256  
                                              reversable defect              564  
non-anginal pain ST-T wave abnormality        fixed defect                   240  
                                              reversable defect              298  
                 normal                       fixed defect                   283  
                                              normal                         233  
                                              reversable defect              288  
typical angina   ST-T wave abnormality        fixed defect                   354  
                                              normal                         315  
                                              reversable defect              353  
                 left ventricular hypertrophy fixed defect                   327  
                                              normal                         318  
                                              reversable defect              205  
                 normal                       fixed defect                   394  
                                              normal                         276  
                                              reversable defect              409

**Сколько значений может принимать каждый из категориальных признаков?**

In [51]:
coun = ['age', 'resting_blood_pressure', 'cholesterol', 'max_heart_rate_achieved', 'st_depression', 'num_major_vessels', 'target']
data_cat = data.drop(coun, axis = 1)

data_cat.nunique()

sex                        2
chest_pain_type            3
fasting_blood_sugar        2
rest_ecg                   3
exercise_induced_angina    2
st_slope                   2
thalassemia                3
dtype: int64

**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [116]:
cat1 = data_cat.describe(include='all')
cat_freq = cat1.iloc[3].astype('int')
cat_count = cat1.iloc[0].astype('int')
cat_freq/cat_count*100

sex                        68.316832
chest_pain_type            63.696370
fasting_blood_sugar        85.148515
rest_ecg                   50.165017
exercise_induced_angina    67.326733
st_slope                   53.135314
thalassemia                54.785479
dtype: float64